### PR TITLE
Add SHA256 utility implementation

### PR DESCRIPTION
**Pull request by emersonknapp**
To be used for REP-2011 type version hashing.
Simple sha256 implementation to generate a 256-byte message digest for any data. Implementation originally copied from Brad Conte https://github.com/B-Con/crypto-algorithms, with modifications to fit code style and fix compiler warnings.

Prereq for ros2/rcl#1027
Part of https://github.com/ros2/ros2/issues/1159

### CODE REVIEW

File: src/sha256.c
Comment (clalancette) [inline src/sha256.c:18]: This is not needed in the C file, only in the header file.

File: src/sha256.c
Comment (clalancette) [inline src/sha256.c:32]: For what it is worth, I think for most of these we could use an inline function rather than a `#define`.  That would give us better type checking, at least.

File: src/sha256.c
Comment (clalancette) [inline src/sha256.c:47]: According to our style, this is over-indented; it should just be 2 spaces.  Same with all of the rest of the code below.

File: src/sha256.c
Comment (clalancette) [inline src/sha256.c:49]: According to our style, this loop should use brackets.  Same for other loops below.

**Review by clalancette — CHANGES_REQUESTED** [review]
Besides the things I've pointed out inline, we should also add tests for this implementation.

All of that said, the idea here is good.

**Comment by emersonknapp** [comment]
Thanks for the comments. I hadn't finished doing code style yet, just wanted to have this open as draft for the rep2011 discussion today.

Agreed, tests necessary.

File: src/sha256.c
Comment (emersonknapp) [reply src/sha256.c]: Agreed - done as `static inline uint32_t` for all

**Review by fujitatomoya — COMMENTED** [review]
@emersonknapp @clalancette quick question, why we are not using openssl to avoid re-implementing the function in rcutils? i was thinking that would be more cost effective for maintenance? Do we want to avoid the dependency on that intentionally? just a question.

**Comment by emersonknapp** [comment]
Yes, the motivation is that we want to use this function in RCL, which does not have a direct dependency on OpenSSL. ros2-security does have the OpenSSL dependency, but not all installations will enable that feature, but type version hashing will be built into rcl. 

@wjwwood may also have more input on the reason

File: src/sha256.c
Comment (clalancette) [inline src/sha256.c:65]:
```suggestion
static void sha256_transform(rcutils_sha256_ctx_t * ctx, const uint8_t data[])
```

(it is not part of the public API)

File: src/sha256.c
Comment (clalancette) [inline src/sha256.c:21]: The `32 - b` is slightly suspect.  What if `b` is 255?  We would then end up shifting left with a negative number, which I think is undefined behavior.  I guess given how we use this function it won't happen, but maybe we should add an assert or something?  What do you think?

File: src/sha256.c
Comment (clalancette) [reply src/sha256.c:65]: Additionally, do we even need the second argument?  We already have `ctx`, and all uses of this function pass `ctx->data` as the second argument.  But we could just dereference that within this function itself.

File: src/sha256.c
Comment (clalancette) [inline src/sha256.c:134]: I'm not sure it is worth it, but we could probably optimize this.  In particular, we could `memcpy` 64 bytes at a time into `ctx->data` (dealing with the special case of the first block and the last block), calling `sha256_transform` on each 64-byte block we complete.

I'll leave it up to you whether you think that improvement is worth it and you want to try it.  This is not a blocking comment from me.

File: src/sha256.c
Comment (clalancette) [inline src/sha256.c:146]: Same comment here about optimizing; this would probably be faster as a `memset`.

File: src/sha256.c
Comment (clalancette) [inline src/sha256.c:151]: Ditto.

File: include/rcutils/sha256.h
Comment (clalancette) [inline include/rcutils/sha256.h:86]: I know the documentation says that `hash` is an out parameter, but it might be nice to encode that in the name:

```suggestion
void rcutils_sha256_final(rcutils_sha256_ctx_t * ctx, uint8_t output_hash[RCUTILS_SHA256_BLOCK_SIZE]);
```

File: test/test_sha256.cpp
Comment (clalancette) [inline test/test_sha256.cpp:22]:
```suggestion
  uint8_t expected_hash1[RCUTILS_SHA256_BLOCK_SIZE] = {
```

File: test/test_sha256.cpp
Comment (clalancette) [inline test/test_sha256.cpp:41]:
```suggestion
  uint8_t expected_hash2[RCUTILS_SHA256_BLOCK_SIZE] = {
```

File: test/test_sha256.cpp
Comment (clalancette) [inline test/test_sha256.cpp:60]:
```suggestion
  uint8_t expected_hash[RCUTILS_SHA256_BLOCK_SIZE] = {
```

**Review by clalancette — CHANGES_REQUESTED** [review]
I've left a few comments, but they are all pretty minor.  Overall this looks pretty good to me.

File: src/sha256.c
Comment (emersonknapp) [reply src/sha256.c:21]: Yes, this is really a `uint5_t` expected input range. I think an assert is fine as long as we're optimizing them out in release builds.

File: src/sha256.c
Comment (emersonknapp) [reply src/sha256.c]: I gave block-copy implementation a go. The tests still pass so it seems all right. I don't have a performance comparison but maybe it looks better this way.

File: src/sha256.c
Comment (emersonknapp) [reply src/sha256.c]: Done

**Review by clalancette — APPROVED** [review]
Thanks for iterating.  This looks good to me with green CI.

**Comment by emersonknapp** [comment]
Gist: https://gist.githubusercontent.com/emersonknapp/e2b18a4fab8fcd32f47ceb057768b8e8/raw/fddd335b1bc03a0c5381e82c3b3989fae80e2d5f/ros2.repos
BUILD args:  --packages-above-and-dependencies rcutils
TEST args:  --packages-above rcutils
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/11402
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17994/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12524/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18631/)

**Comment by emersonknapp** [comment]
unrelated rclcpp test failed, rerunning Linux:
* Linux [Build Status](https://ci.ros2.org/job/ci_linux/17998/)

**Comment by clalancette** [comment]
You can ignore that failure; it is fixed in https://github.com/ros2/rclcpp/pull/2092 .

**Comment by emersonknapp** [comment]
OK then :+1:, I don't have merge access on this repo so that's up to you now! Thanks for the reviews

**Pull request merged**
